wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs

```diff
@@ -0,0 +1,194 @@
+             return ((InternalStorage us StorageReleased), ())
+
+
+{-------------------------------------------------------------------------------
+  Destroying a keystore
+-------------------------------------------------------------------------------}
+
+-- | Destroys a 'Keystore'.
+-- Completely obliterate the keystore from disk, with all its secrets.
+-- This operation cannot be reverted.
+-- This is a very destructive option that most of the time you
+-- probably don't want, as it will effectively leave you with an empty
+-- 'MVar' and your next call to an operation on the keystore would block
+-- indefinitely.
+-- This has still its use in some teardowns or tests.
+-- Note that this operation will always succeed. Use with care.
+destroyKeystore :: MonadIO m => Keystore -> m ()
+destroyKeystore (Keystore ks) = liftIO $ do
+    (InternalStorage us _st) <- takeMVar ks
+    writeUserSecretRelease us
+    removeFile (getUSPath us)
+
+{-------------------------------------------------------------------------------
+  Inserting things inside a keystore
+-------------------------------------------------------------------------------}
+
+-- | Insert a new 'EncryptedSecretKey' indexed by the input 'WalletId'.
+insert :: MonadIO m
+       => WalletId
+       -> EncryptedSecretKey
+       -> Keystore
+       -> m ()
+insert _walletId esk (Keystore ks) =
+    liftIO $ modifyMVar_ ks $ \(InternalStorage us st) -> do
+        return $ if view usKeys us `contains` esk
+                     then InternalStorage us st
+                     else InternalStorage (us & over usKeys (esk :)) st
+    where
+      -- Comparator taken from the old code which needs to hash
+      -- all the 'EncryptedSecretKey' in order to compare them.
+      contains :: [EncryptedSecretKey] -> EncryptedSecretKey -> Bool
+      contains ls k = hash k `elem` map hash ls
+
+{-------------------------------------------------------------------------------
+  Looking up things inside a keystore
+-------------------------------------------------------------------------------}
+
+-- | Lookup an 'EncryptedSecretKey' associated to the input 'HdRootId'.
+lookup :: MonadIO m
+       => WalletId
+       -> Keystore
+       -> m (Maybe EncryptedSecretKey)
+lookup (WalletIdHdRnd walletId) (Keystore ks) =
+    liftIO $ withMVar ks $ \(InternalStorage us st) ->
+        return $ case st of
+            StorageInitialised ->
+                Data.List.find (\k -> eskToHdRootId k == walletId) (us ^. usKeys)
+            StorageReleased      -> Nothing
+
+{-------------------------------------------------------------------------------
+  Deleting things from the keystore
+-------------------------------------------------------------------------------}
+delete :: MonadIO m
+       => WalletId
+       -> Keystore
+       -> m ()
+delete walletId (Keystore ks) = do
+    mbEsk <- lookup walletId (Keystore ks)
+    liftIO $ modifyMVar_ ks $ \(InternalStorage us st) ->
```

Comment:
Let's avoid the race condition here. Split `lookup` into a top-level function and an internal pure one, then here use the internal one with a single `modifyMVar`.